### ISSUE TITLE
Slow Hessian

### ISSUE THREAD

**Issue by amehrjou** [tensorflow/tensorflow#11894]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: I have written custom code
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 4.4.0-47-generic #68-Ubuntu and Also macOS Sierra 10.12.3 (CPU only) 
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: 1
- **Python version**: 3.6
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: cuda/8.0 and cudnn/5.1
- **GPU model and memory**: 12GB, memory:100GB
- **Exact command to reproduce**: tf.hessians(ys,xs)

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
The command tf.hessians(ys, xs) is used to add nodes to the graph in order to compute hessians of ys with respect to xs where both ys and xs are list of tensors. Currently computing hessian with respect to a vector is not possible. The trick is to unstack the input tensor (x) into a list of one dimensional tensors (xs) and compute hessian with respect to each of them separately then stack them together. Tensorflow gets stuck in the phase of graph construction even when the dimension of x (length of list xs) is about one hundred. In the implementation of Hessian in gradients_impl.py the second derivative is implemented as the derivative of partial derivative with respect to each member of xs. I guess the slowness of graph construction is due to this line:

_hess = [gradients(_gradient, x, **kwargs)[0] for _gradient in _gradients]

which may add several unnecessary intermediate nodes with overlapping functionality to the graph due to the for loop. Is there any way other than looping over input dimensions that efficiently constructs the graph in a reasonable time?

### Source code / logs
This source code can simulate the problem:
import tensorflow as tf
import numpy as np
in_dimension = 256
x = tf.placeholder(tf.float32, shape=(1, in_dimension))
x_list = tf.unstack(x, axis=1)
xx= tf.stack(x_list, axis=1)
y = tf.pow(xx,3)
hess = tf.hessians(y, x_list)
sess = tf.Session()
print(sess.run(hess, feed_dict={x : np.random.normal(0, 1, size=(1, in_dimension))}))

**Comment by aselle** [comment]
@alextp, it seems like slowness is inherent with the current approach to computing Hessians of non-scalars. I don't know of anything else to do. Could eager mode help with this at all?

**Comment by yaroslavvb** [comment]
@amehrjou one issue is that current`tf.gradients` implementation is quite slow. The runtime grows linearly with size of the graph even when size of the subgraph to differentiate or the output stays constant. https://github.com/tensorflow/tensorflow/issues/9901

**Comment by amehrjou** [comment]
@yaroslavvb Exactly. I also realized that the size of the graph becomes very large when `tf.gradients` is applied for the second time. For example the first `tf.gradients` added around 100 nodes to a graph with 80 nodes. But the second derivative surprisingly adds 1000 nodes to the graph for each dimension of the vector with respect to which derivates are computed. Is there any chance to compute Hessian diagonal in a feasible way?

**Comment by yaroslavvb** [comment]
Oh you want just the diagonal of Hessian? For piecewise linear neural network (ie, ReLU activations) computing diagonal can be done cheaply (ie, for square loss it's the product of corresponding activation and backprop variances). In a general network, there's no fast way (http://www.cs.utoronto.ca/~ilya/pubs/2012/CurvProp.pdf)

**Comment by amehrjou** [comment]
@yaroslavvb I want the diagonal of Hessian but specifically for a smooth network and I want it with respect to inputs not parameters. Second derivative of a piecewise linear function with respect to its input is trivially zero and just for the record Tensorflow returns `None` even though it must return zero.

**Comment by yaroslavvb** [comment]
`None` result in `tf.gradients` means there's no path between input and output in the computation graph

**Comment by amehrjou** [comment]
@yaroslavvb Yes. I replace `None` values with zero. Thank you for the pointer to the Curve Propagation paper too.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.
